### PR TITLE
Add cufile I/O engine

### PR DESCRIPTION
**Pull request by bsmith94**
Please review. I hope that the code adheres to the patterns I observed in the other I/O engines, and am sure there are things that need to be corrected.

All feedback is welcome.

As of GPUDirect Storage 0.9.0, the configure invocation is:

```
$ CFLAGS="-I/usr/local/cuda/include -I/usr/local/cuda/lib64" LDFLAGS="-L/usr/local/cuda/lib64" ./configure --enable-cufile
```

GPUDirect Storage (GDS) is obtained from here: https://developer.nvidia.com/gpudirect-storage

GDS requires a Tesla or better GPU, x86_64 CPU, Ubuntu OS 18 or 20, MOFED 5.1 and Mellanox HCA >= ConnectX-5.

----

The cufile I/O engine uses NVIDIA GPUDirect Storage (GDS) cuFile API to perform
I/O directly against GPU buffers via nvidia-fs and a GDS-supported filesystem.

'configure --enable-cufile' enables the cufile engine.

CFLAGS must specify the location of cuFile headers, e.g.
CFLAGS="-I /usr/local/gds/lib"

LDFLAGS must specify the location of cuFile libraries .e.g
LDFLAGS="-L /usr/local/gds/lib"

cuFile will eventually be part of CUDA, and will then not require
the special CFLAGS and LDFLAGS settings.

cufile engine options:

    cufile_gpu_ids : colon-separated list of GPU-indexes to use, per subjob.
       If the quantity of IDs in the list is less than the quantity of subjobs,
       the last ID in the list will be used for remaining subjobs.

    cufile_posixio : if 1, the cufile engine will instead use POSIX I/O with
       CPU buffers and perform cudaMemcpy() as necessary. This provides
       relative performance for a CUDA application the uses POSIX I/O instead
       of the cuFile API.

Note that enabling validation necessitates cudaMemcpy() to populate
the GPU buffer on a write and populate the CPU buffer on a read.

### CODE REVIEW

File: engines/cufile.c [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+		cudaError_t res = fn;										\
+		if (res != cudaSuccess) {									\
+			const char *str = cudaGetErrorName(res);				\
+			log_err("cuda runtime api call failed %s:%d : %s\n",	\
+					#fn, __LINE__, str);							\
+			rc = -1;												\
+		} else {													\
+			rc = 0;													\
+		}															\
+	} while(0)
```
Comment (sitsofe) [inline engines/cufile.c:95]: I'll assume GitHub has done a poor job of showing how these line up :-)

File: engines/cufile.c [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+	char* cur = NULL;
+	char* pos = NULL;
+	CUfileError_t status;
+	int gpu_id = 0;
+	int i;
+	int initialized;
+	int rc;
+
+	pthread_mutex_lock(&running_lock);
+	if (running == 0) {
```
Comment (sitsofe) [inline engines/cufile.c:120]: If this is running in a separate process to other versions of this engine will that be an issue?

File: engines/cufile.c [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+	running++;
+	initialized = cufile_initialized;
+	pthread_mutex_unlock(&running_lock);
+
+	if (!o->use_posixio && !initialized) {
+		return 1;
+	}
+
+	if (o->gpu_ids) {
+		for (i = 0, cur = strtok_r(o->gpu_ids, GPU_ID_SEP, &pos);
```
Comment (sitsofe) [inline engines/cufile.c:142]: Ah string splitting in C is always nasty. However I would feel better if we were to `strsep`...

File: engines/cufile.c [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+	}
+
+	return 0;
+}
+
+static inline int fio_cufile_pre_write(struct thread_data *td,
+									   struct cufile_options *o,
+									   struct io_u *io_u,
+									   size_t gpu_offset)
+{
```
Comment (sitsofe) [inline engines/cufile.c:163]: Again I can tell if this will really line up in real life...

File: engines/cufile.c [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+		}
+	} else {
+		o->junk_buf = NULL;
+	}
+
+	dprint(FD_MEM, "Alloc %zu for GPU %d\n", total_mem, o->my_gpu_id);
+	check_cudaruntimecall(cudaMalloc(&o->cu_mem_ptr, total_mem), rc);
+	if (rc != 0) {
+		goto exit_error;
+	}
```
Comment (sitsofe) [inline engines/cufile.c:489]: House style is that single line conditional bodies without an else don't get braces.

File: engines/cufile.c [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+	check_cudaruntimecall(cudaMemset(o->cu_mem_ptr, 0xab, total_mem), rc);
+	if (rc != 0) {
+		goto exit_error;
+	}
```
Comment (sitsofe) [inline engines/cufile.c:493]: As above.

File: engines/cufile.c [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+
+	if (!o->use_posixio) {
+		status = cuFileBufRegister(o->cu_mem_ptr, total_mem, 0);
+		if (status.err != CU_FILE_SUCCESS) {
+			log_err("cuFileBufRegister: %d:%s\n", status.err,
+					cuFileGetErrorString(status));
```
Comment (sitsofe) [inline engines/cufile.c:499]: I'd line up the following line with the open bracket.

File: engines/cufile.c [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+	.iomem_free          = fio_cufile_iomem_free,
+	.cleanup             = fio_cufile_cleanup,
+	.flags               = FIO_SYNCIO,
+	.options             = options,
+	.option_struct_size  = sizeof(struct cufile_options)
+};
+
+void fio_init fio_cufile_register(void)
+{
+	const char* diskless = getenv("FIO_CUFILE_DISKLESSIO");
```
Comment (sitsofe) [inline engines/cufile.c:576]: Hmm this is different. We don't typically pass parameters via environment variable... What forced you to do so here?

**Review by sitsofe — CHANGES_REQUESTED** [review]
- Please prefix your commit with something like `ioengine: [...]`
- Please add your Signed-off-by
- We're missing documentation for this ioengine in the HOWTO and fio.1
- We're missing an example jobfile
- Won't compilation flags depend entirely on where CUDA was installed? If it is in the usual place it would hopefully be auto-detected... 

Out of curiosity how come we need another CUDA-based engine? What's this one testing differently?

File: engines/cufile.c (hunk 1 above)
Comment (bsmith94) [reply engines/cufile.c:95]: github is using a different tab width than my editor. I'll change the whitespace to the line continuation backslash to use spaces. That should tame it.

File: engines/cufile.c (hunk 2 above)
Comment (bsmith94) [reply engines/cufile.c:120]: Sorry, I'm not sure that I understand the question. This is designed to operate in both multi-process and multi-threaded mode. If you are referring to client/server operation, there is no explicit data transfer implemented by this engine. Everything happens through the filesystem. So, I believe the answer is: "No, there should not be any issue."

File: engines/cufile.c (hunk 3 above)
Comment (bsmith94) [reply engines/cufile.c:142]: Will do.

File: engines/cufile.c (hunk 4 above)
Comment (bsmith94) [reply engines/cufile.c:163]: I will adjust my editor settings and reformat the code.

File: engines/cufile.c (hunk 5 above)
Comment (bsmith94) [reply engines/cufile.c:489]: OK, will do.

File: engines/cufile.c (hunk 6 above)
Comment (bsmith94) [reply engines/cufile.c:493]: OK, will do.

File: engines/cufile.c (hunk 7 above)
Comment (bsmith94) [reply engines/cufile.c:499]: OK, will do.

File: engines/cufile.c (hunk 8 above)
Comment (bsmith94) [reply engines/cufile.c:576]: Agreed, that is out of the ordinary. The reason for using an environment variable is because the underlying filesystem may or may not be diskless. Using an fio_option to set the flag didn't appear to work. I was also unclear on the total ramifications of setting FIO_DISKLESSIO. I see it used in engines/cpu.c and engines/rdma.c, which makes perfect sense, but also in engines/glusterfs_sync.c, and that made things murkier. My current take is that FIO_DISKLESSIO should be set if the engine *should never* access local disk.

Long story short, I don't believe that FIO_DISKLESSIO needs to be set and will remove that code.

**Comment by bsmith94** [comment]
> Won't compilation flags depend entirely on where CUDA was installed? 

Yes. The existing work to support `configure --enable-cuda` also requires CFLAGS and LDFLAGS to add the CUDA directories and that doesn't appear to be documented anywhere. I added that note as a point of reference.

> Out of curiosity how come we need another CUDA-based engine? What's this one testing differently?

The existing CUDA code tests GPUDirect through use of `--iomem=cudamalloc` and `--ioengine=rdma`. That tests I/O between two GPUs via an RDMA interface.

The cufile I/O engine tests performance of nvidia-fs I/O with GPU buffers, RDMA and a GPUDirectStorage-supported filesystem (e.g. ext4 on nvme, nfsordma, lustre). This is an entirely different use case than what is implemented by  engines/rdma.c. 

`--iomem=cudamalloc` is not used for cufile because we still need the malloc'd io buffers in order to implement verify. cufile does its own allocation of GPU buffers, with the same capacity of what is malloc'd for io buffers.

The option `--cufile_posixio` allows comparison of I/O between "CUDA using nvidia-fs" and "CUDA using POSIX I/O". The default behavior of `--cufile_posixio=0` uses nvidia-fs to do I/O directly with the GPU buffers--the data is never copied between the GPU and RAM (unless verify is enabled). In the POSIX I/O case, I/O is performed in a RAM buffer and that has to be synced up with the GPU buffer via cudaMemcpy.

**Comment by sitsofe** [comment]
@bsmith94 Could you squash the new changes into the original commit and change the original's title/body?

**Comment by bsmith94** [comment]
@sitsofe I did not realize that mentioning the pull request number in my commit to a different branch would auto-update the pull request. Sorry about that. I still have a few things to address, and the plan is to squash everything into a single commit.

File: engines/libcufile.c [hunk 9]
```diff
@@ -0,0 +1,624 @@
...
+	pthread_mutex_unlock(&running_lock);
+
+	if (o->cuda_io == IO_CUFILE && !initialized)
+		return 1;
+
+	o->my_gpu_id = fio_libcufile_find_gpu_id(td);
+	if (o->my_gpu_id < 0)
+		return 1;
+
+	dprint(FD_MEM, "Subjob %d use GPU %d\n", td->subjob_number, o->my_gpu_id);
```
Comment (sitsofe) [inline engines/libcufile.c:190]: use -> uses

File: examples/libcufile-cufile.fio [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+rw=randwrite
+
+[verify]
+rw=write
+verify=md5
+
+[randverify]
+rw=randwrite
+verify=md5
+
```
Comment (sitsofe) [inline examples/libcufile-cufile.fio:43]: Maybe trim the trailing newline

File: examples/libcufile-posix.fio [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+rw=randwrite
+
+[verify]
+rw=write
+verify=md5
+
+[randverify]
+rw=randwrite
+verify=md5
+
```
Comment (sitsofe) [inline examples/libcufile-posix.fio:42]: Maybe trim the trailing newline

**Review by sitsofe — APPROVED** [review]
Some trivial comments but LGTM! @axboe this look pullable...

File: engines/libcufile.c (hunk 9 above)
Comment (bsmith94) [reply engines/libcufile.c:190]: ok

File: examples/libcufile-cufile.fio (hunk 10 above)
Comment (bsmith94) [reply examples/libcufile-cufile.fio:43]: ok. I think the sections should be stonewalled. Thoughts?

File: examples/libcufile-posix.fio (hunk 11 above)
Comment (bsmith94) [reply examples/libcufile-posix.fio:42]: ok

Comment (bsmith94) [reply examples/libcufile-posix.fio:42]: Should I squash the changes into one commit or push a new commit?

Comment (bsmith94) [reply examples/libcufile-posix.fio]: I squashed.

**Comment by bsmith94** [comment]
Thanks for helping me get this pulled.
